File: mobile/composeApp/src/commonMain/kotlin/org/androidbible/data/bintex/Yes2Reader.kt
package org.androidbible.data.bintex

/**
 * Kotlin port of yuku.alkitab.yes2.Yes2Reader — reads YES2 Bible files.
 *
 * YES2 header: 0x98 0x58 0x0D 0x0A 0x00 0x5D 0xE0 0x02
 * Section index starts at offset 12.
 * Sections: versionInfo, booksInfo, text, pericopes, xrefs, footnotes.
 */
class Yes2Reader(private val data: ByteArray) {

    private val sectionEntries = LinkedHashMap<String, SectionEntry>()
    private var sectionDataStartOffset: Int = 0

    // Cached data
    private var cachedVersionInfo: Yes2VersionInfo? = null
    private var cachedBooksInfo: List<Yes2Book>? = null
    private var cachedTextData: ByteArray? = null

    init {
        parseHeader()
        parseSectionIndex()
    }

    // -- Header & Index parsing --

    private fun parseHeader() {
        check(data.size >= HEADER.size) { "YES2: File too small for header" }
        for (i in HEADER.indices) {
            check(data[i] == HEADER[i]) { "YES2: Invalid header at byte $i" }
        }
    }

    private fun parseSectionIndex() {
        val br = BintexReader(data, SECTION_INDEX_OFFSET)

        val version = br.readUint8()
        check(version == 1) { "YES2: Unsupported section index version: $version" }

        val sectionCount = br.readInt()
        repeat(sectionCount) {
            val nameLen = br.readUint8()
            val nameBytes = br.readRaw(nameLen)
            // Section names are ASCII/Latin-1
            val name = String(CharArray(nameBytes.size) { (nameBytes[it].toInt() and 0xFF).toChar() })

            val offset = br.readInt()
            val attributesSize = br.readInt()
            val contentSize = br.readInt()
            br.skip(4) // reserved

            sectionEntries[name] = SectionEntry(name, offset, attributesSize, contentSize)
        }

        sectionDataStartOffset = br.pos
    }

    val sectionNames: List<String> get() = sectionEntries.keys.toList()

    // -- Section offsets --

    private fun getSectionAttributesOffset(name: String): Int? {
        val entry = sectionEntries[name] ?: return null
        return sectionDataStartOffset + entry.offset
    }

    private fun getSectionContentOffset(name: String): Int? {
        val entry = sectionEntries[name] ?: return null
        return sectionDataStartOffset + entry.offset + entry.attributesSize
    }

    private fun readSectionAttributes(name: String): Map<String, Any?>? {
        val offset = getSectionAttributesOffset(name) ?: return null
        val entry = sectionEntries[name]!!
        if (entry.attributesSize == 0) return null
        val br = BintexReader(data, offset)
        return br.readValueSimpleMap()
    }

    // -- Version Info --

    fun getVersionInfo(): Yes2VersionInfo {
        cachedVersionInfo?.let { return it }

        val offset = getSectionContentOffset("versionInfo")
            ?: error("YES2: versionInfo section not found")

        val br = BintexReader(data, offset)
        val map = br.readValueSimpleMap()

        val info = Yes2VersionInfo(
            shortName = map["shortName"] as? String,
            longName = map["longName"] as? String,
            description = map["description"] as? String,
            locale = map["locale"] as? String,
            bookCount = (map["book_count"] as? Int) ?: 0,
            hasPericopes = ((map["hasPericopes"] as? Int) ?: 0) != 0,
            textEncoding = (map["textEncoding"] as? Int) ?: 2,
        )
        cachedVersionInfo = info
        return info
    }

    // -- Books Info --

    fun getBooksInfo(): List<Yes2Book> {
        cachedBooksInfo?.let { return it }

        val offset = getSectionContentOffset("booksInfo")
            ?: error("YES2: booksInfo section not found")

        val br = BintexReader(data, offset)
        val bookCount = br.readInt()

        val books = ArrayList<Yes2Book>(bookCount)
        repeat(bookCount) {
            val map = br.readValueSimpleMap()
            books.add(
                Yes2Book(
                    bookId = (map["bookId"] as? Int) ?: -1,
                    shortName = map["shortName"] as? String,
                    abbreviation = map["abbreviation"] as? String,
                    offset = (map["offset"] as? Int) ?: 0,
                    chapterCount = (map["chapter_count"] as? Int) ?: 0,
                    verseCounts = (map["verse_counts"] as? IntArray) ?: intArrayOf(),
                    chapterOffsets = (map["chapter_offsets"] as? IntArray) ?: intArrayOf(),
                )
            )
        }

        cachedBooksInfo = books
        return books
    }

    /**
     * Find a book by bookId (1-based, e.g. Gen=1, Rev=66).
     */
    fun findBookByBookId(bookId: Int): Pair<Int, Yes2Book>? {
        val books = getBooksInfo()
        val index = books.indexOfFirst { it.bookId == bookId }
        return if (index >= 0) index to books[index] else null
    }

    // -- Text section --

    /**
     * Load verse texts for a specific book index and 1-based chapter.
     * Returns a list of verse strings (0-indexed within the returned list).
     */
    fun loadVerseText(bookIndex: Int, chapter1: Int): List<String> {
        val books = getBooksInfo()
        check(bookIndex in books.indices) { "YES2: Book index $bookIndex not found" }

        val book = books[bookIndex]
        check(chapter1 in 1..book.chapterCount) {
            "YES2: Chapter $chapter1 out of range for book $bookIndex"
        }

        val versionInfo = getVersionInfo()
        val textData = getDecompressedTextData()

        val contentOffset = book.offset + book.chapterOffsets[chapter1 - 1]
        val verseCount = book.verseCounts[chapter1 - 1]

        val br = BintexReader(textData, contentOffset)
        return readVerses(br, verseCount, versionInfo.textEncoding)
    }

    /**
     * Load a single verse by bookIndex, 1-based chapter, and 1-based verse.
     */
    fun loadSingleVerse(bookIndex: Int, chapter1: Int, verse1: Int): String? {
        val books = getBooksInfo()
        if (bookIndex !in books.indices) return null

        val book = books[bookIndex]
        if (chapter1 !in 1..book.chapterCount) return null
        if (verse1 < 1 || verse1 > book.verseCounts[chapter1 - 1]) return null

        val versionInfo = getVersionInfo()
        val textData = getDecompressedTextData()

        val contentOffset = book.offset + book.chapterOffsets[chapter1 - 1]
        val br = BintexReader(textData, contentOffset)

        // Skip to the target verse
        repeat(verse1 - 1) {
            val len = br.readVarUint()
            br.skip(len)
        }

        val len = br.readVarUint()
        val raw = br.readRaw(len)
        return decodeVerseBytes(raw, versionInfo.textEncoding)
    }

    private fun getDecompressedTextData(): ByteArray {
        cachedTextData?.let { return it }

        val sectionAttributes = readSectionAttributes("text")
        val contentOffset = getSectionContentOffset("text")
            ?: error("YES2: text section not found")

        val textData = decompressSection("text", sectionAttributes, contentOffset)
        cachedTextData = textData
        return textData
    }

    private fun decompressSection(
        sectionName: String,
        sectionAttributes: Map<String, Any?>?,
        contentOffset: Int,
    ): ByteArray {
        val entry = sectionEntries[sectionName]!!

        if (sectionAttributes == null) {
            return data.copyOfRange(contentOffset, contentOffset + entry.contentSize)
        }

        val compressionName = sectionAttributes["compression.name"] as? String
        if (compressionName == null) {
            return data.copyOfRange(contentOffset, contentOffset + entry.contentSize)
        }

        check(compressionName == "snappy-blocks") { "YES2: Unsupported compression: $compressionName" }

        @Suppress("UNCHECKED_CAST")
        val compressionInfo = sectionAttributes["compression.info"] as? Map<String, Any?> ?: emptyMap()
        val blockSize = (compressionInfo["block_size"] as? Int) ?: 0
        val compressedBlockSizes = (compressionInfo["compressed_block_sizes"] as? IntArray) ?: intArrayOf()

        check(blockSize > 0 && compressedBlockSizes.isNotEmpty()) {
            "YES2: Invalid snappy-blocks compression info"
        }

        // Compute offsets and decompress each block
        val blocks = ArrayList<ByteArray>(compressedBlockSizes.size)
        var totalSize = 0
        var blockOffset = 0
        for (compSize in compressedBlockSizes) {
            val blockData = data.copyOfRange(contentOffset + blockOffset, contentOffset + blockOffset + compSize)
            val decompressed = SnappyCodec.decompress(blockData)
            blocks.add(decompressed)
            totalSize += decompressed.size
            blockOffset += compSize
        }

        // Concatenate all decompressed blocks
        val result = ByteArray(totalSize)
        var pos = 0
        for (block in blocks) {
            block.copyInto(result, pos)
            pos += block.size
        }
        return result
    }

    private fun readVerses(br: BintexReader, verseCount: Int, textEncoding: Int): List<String> {
        val verses = ArrayList<String>(verseCount)
        repeat(verseCount) {
            val len = br.readVarUint()
            val raw = br.readRaw(len)
            verses.add(decodeVerseBytes(raw, textEncoding))
        }
        return verses
    }

    private fun decodeVerseBytes(raw: ByteArray, textEncoding: Int): String {
        return if (textEncoding == 1) {
            // Latin-1
            String(CharArray(raw.size) { (raw[it].toInt() and 0xFF).toChar() })
        } else {
            // UTF-8
            raw.decodeToString()
        }
    }

    // -- Pericopes --

    fun loadPericopes(): List<PericopeEntry> {
        val versionInfo = getVersionInfo()
        if (!versionInfo.hasPericopes) return emptyList()

        val sectionAttributes = readSectionAttributes("pericopes")
        val contentOffset = getSectionContentOffset("pericopes") ?: return emptyList()

        val sectionData = decompressSection("pericopes", sectionAttributes, contentOffset)
        val br = BintexReader(sectionData)

        val version = br.readUint8()
        check(version == 2 || version == 3) { "YES2: Unsupported pericope version: $version" }

        /* indexSize = */ br.readInt()
        val entryCount = br.readInt()

        val aris = IntArray(entryCount)
        val offsets = IntArray(entryCount)

        if (version == 2) {
            repeat(entryCount) { i ->
                aris[i] = br.readInt()
                offsets[i] = br.readInt()
            }
        } else {
            // version 3 — delta-encoded
            var lastAri = 0
            var lastOffset = 0
            repeat(entryCount) { i ->
                val dataAri = br.readUint16()
                val ari = if ((dataAri and 0x8000) == 0) {
                    (dataAri shl 16) or br.readUint16()
                } else {
                    lastAri + (dataAri and 0x7FFF)
                }
                aris[i] = ari
                lastAri = ari

                val dataOffset = br.readUint16()
                val offset = lastOffset + dataOffset
                offsets[i] = offset
                lastOffset = offset
            }
        }

        val dataStartPos = br.pos
        val result = ArrayList<PericopeEntry>(entryCount)

        repeat(entryCount) { i ->
            br.seek(dataStartPos + offsets[i])
            /* blockVersion = */ br.readUint8()
            val title = br.readValueString() ?: ""
            val parallelCount = br.readUint8()
            val parallels = ArrayList<String>(parallelCount)
            repeat(parallelCount) {
                br.readValueString()?.let { parallels.add(it) }
            }
            result.add(PericopeEntry(aris[i], title, parallels))
        }

        return result
    }

    // -- Xrefs & Footnotes --

    /**
     * Load cross-references. Returns map of ARIF → content string.
     */
    fun loadXrefs(): Map<Int, String> = loadArifSection("xrefs")

    /**
     * Load footnotes. Returns map of ARIF → content string.
     */
    fun loadFootnotes(): Map<Int, String> = loadArifSection("footnotes")

    private fun loadArifSection(sectionName: String): Map<Int, String> {
        val sectionAttributes = readSectionAttributes(sectionName)
        val contentOffset = getSectionContentOffset(sectionName) ?: return emptyMap()

        val sectionData = decompressSection(sectionName, sectionAttributes, contentOffset)
        val br = BintexReader(sectionData)

        val version = br.readUint8()
        check(version == 1) { "YES2: Unsupported $sectionName version: $version" }

        val entryCount = br.readInt()
        val arifs = IntArray(entryCount) { br.readInt() }
        val offsets = IntArray(entryCount) { br.readInt() }

        val dataStartPos = br.pos
        val result = LinkedHashMap<Int, String>(entryCount)

        repeat(entryCount) { i ->
            br.seek(dataStartPos + offsets[i])
            val content = br.readValueString()
            if (content != null) {
                result[arifs[i]] = content
            }
        }

        return result
    }

    companion object {
        private val HEADER = byteArrayOf(
            0x98.toByte(), 0x58, 0x0D, 0x0A, 0x00, 0x5D, 0xE0.toByte(), 0x02
        )
        private const val SECTION_INDEX_OFFSET = 12

        fun fromByteArray(data: ByteArray): Yes2Reader = Yes2Reader(data)
    }
}

File: backend/app/Services/Bintex/Yes2Reader.php
<?php

declare(strict_types=1);

namespace App\Services\Bintex;

use RuntimeException;

/**
 * PHP port of yuku.alkitab.yes2.Yes2Reader — reads YES2 Bible files.
 *
 * YES2 header: 0x98 0x58 0x0D 0x0A 0x00 0x5D 0xE0 0x02
 * Section index starts at offset 12.
 * Sections: versionInfo, booksInfo, text, pericopes, xrefs, footnotes.
 */
class Yes2Reader
{
    private const HEADER = "\x98\x58\x0D\x0A\x00\x5D\xE0\x02";
    private const HEADER_LENGTH = 8;
    private const SECTION_INDEX_OFFSET = 12;

    private string $data;

    /** @var array<string, array{name: string, offset: int, attributes_size: int, content_size: int}> */
    private array $sectionEntries = [];

    private int $sectionDataStartOffset = 0;

    // Cached section data
    private ?array $versionInfo = null;
    /** @var array<int, array<string, mixed>>|null */
    private ?array $booksInfo = null;

    public function __construct(string $data)
    {
        $this->data = $data;
        $this->parseHeader();
        $this->parseSectionIndex();
    }

    public static function fromFile(string $path): self
    {
        $data = file_get_contents($path);
        if ($data === false) {
            throw new RuntimeException("Cannot read file: {$path}");
        }
        return new self($data);
    }

    private function parseHeader(): void
    {
        if (strlen($this->data) < self::HEADER_LENGTH) {
            throw new RuntimeException('YES2: File too small for header');
        }
        $header = substr($this->data, 0, self::HEADER_LENGTH);
        if ($header !== self::HEADER) {
            throw new RuntimeException('YES2: Invalid header');
        }
    }

    private function parseSectionIndex(): void
    {
        $br = new BintexReader($this->data, self::SECTION_INDEX_OFFSET);

        $version = $br->readUint8();
        if ($version !== 1) {
            throw new RuntimeException("YES2: Unsupported section index version: {$version}");
        }

        $sectionCount = $br->readInt();

        for ($i = 0; $i < $sectionCount; $i++) {
            $nameLen = $br->readUint8();
            $nameRaw = $br->readRaw($nameLen);
            $name = mb_convert_encoding($nameRaw, 'UTF-8', 'ISO-8859-1');

            $offset = $br->readInt();
            $attributesSize = $br->readInt();
            $contentSize = $br->readInt();
            $br->skip(4); // reserved

            $this->sectionEntries[$name] = [
                'name' => $name,
                'offset' => $offset,
                'attributes_size' => $attributesSize,
                'content_size' => $contentSize,
            ];
        }

        $this->sectionDataStartOffset = $br->getPos();
    }

    /**
     * Get absolute offset for a section's attributes.
     */
    private function getSectionAttributesOffset(string $name): ?int
    {
        $entry = $this->sectionEntries[$name] ?? null;
        if ($entry === null) {
            return null;
        }
        return $this->sectionDataStartOffset + $entry['offset'];
    }

    /**
     * Get absolute offset for a section's content (after attributes).
     */
    private function getSectionContentOffset(string $name): ?int
    {
        $entry = $this->sectionEntries[$name] ?? null;
        if ($entry === null) {
            return null;
        }
        return $this->sectionDataStartOffset + $entry['offset'] + $entry['attributes_size'];
    }

    /**
     * Read section attributes as a SimpleMap.
     *
     * @return array<string, mixed>|null
     */
    private function readSectionAttributes(string $name): ?array
    {
        $offset = $this->getSectionAttributesOffset($name);
        if ($offset === null) {
            return null;
        }
        $br = new BintexReader($this->data, $offset);
        return $br->readValueSimpleMap();
    }

    /**
     * Get a list of all section names in this file.
     *
     * @return string[]
     */
    public function getSectionNames(): array
    {
        return array_keys($this->sectionEntries);
    }

    // -- Version Info --

    /**
     * @return array{shortName: ?string, longName: ?string, description: ?string, locale: ?string, book_count: int, hasPericopes: int, textEncoding: int}
     */
    public function getVersionInfo(): array
    {
        if ($this->versionInfo !== null) {
            return $this->versionInfo;
        }

        $offset = $this->getSectionContentOffset('versionInfo');
        if ($offset === null) {
            throw new RuntimeException('YES2: versionInfo section not found');
        }

        $br = new BintexReader($this->data, $offset);
        $map = $br->readValueSimpleMap();

        $this->versionInfo = [
            'shortName' => $map['shortName'] ?? null,
            'longName' => $map['longName'] ?? null,
            'description' => $map['description'] ?? null,
            'locale' => $map['locale'] ?? null,
            'book_count' => (int) ($map['book_count'] ?? 0),
            'hasPericopes' => (int) ($map['hasPericopes'] ?? 0),
            'textEncoding' => (int) ($map['textEncoding'] ?? 2),
        ];

        return $this->versionInfo;
    }

    // -- Books Info --

    /**
     * @return array<int, array{bookId: int, shortName: ?string, abbreviation: ?string, offset: int, chapter_count: int, verse_counts: int[], chapter_offsets: int[]}>
     */
    public function getBooksInfo(): array
    {
        if ($this->booksInfo !== null) {
            return $this->booksInfo;
        }

        $offset = $this->getSectionContentOffset('booksInfo');
        if ($offset === null) {
            throw new RuntimeException('YES2: booksInfo section not found');
        }

        $br = new BintexReader($this->data, $offset);
        $bookCount = $br->readInt();

        $books = [];
        for ($i = 0; $i < $bookCount; $i++) {
            $map = $br->readValueSimpleMap();
            $books[] = [
                'bookId' => (int) ($map['bookId'] ?? -1),
                'shortName' => $map['shortName'] ?? null,
                'abbreviation' => $map['abbreviation'] ?? null,
                'offset' => (int) ($map['offset'] ?? 0),
                'chapter_count' => (int) ($map['chapter_count'] ?? 0),
                'verse_counts' => $map['verse_counts'] ?? [],
                'chapter_offsets' => $map['chapter_offsets'] ?? [],
            ];
        }

        $this->booksInfo = $books;
        return $this->booksInfo;
    }

    // -- Text section --

    /**
     * Load verse texts for a specific book and chapter.
     *
     * @return string[] Array of verse strings
     */
    public function loadVerseText(int $bookIndex, int $chapter1): array
    {
        $books = $this->getBooksInfo();
        if (!isset($books[$bookIndex])) {
            throw new RuntimeException("YES2: Book index {$bookIndex} not found");
        }

        $book = $books[$bookIndex];
        if ($chapter1 <= 0 || $chapter1 > $book['chapter_count']) {
            throw new RuntimeException("YES2: Chapter {$chapter1} out of range for book {$bookIndex}");
        }

        $versionInfo = $this->getVersionInfo();
        $textEncoding = $versionInfo['textEncoding'];

        // Section attributes for text section
        $sectionAttributes = $this->readSectionAttributes('text');
        $sectionContentOffset = $this->getSectionContentOffset('text');
        if ($sectionContentOffset === null) {
            throw new RuntimeException('YES2: text section not found');
        }

        // Compute content offset for this chapter within the text section
        $contentOffset = $book['offset'] + $book['chapter_offsets'][$chapter1 - 1];
        $verseCount = $book['verse_counts'][$chapter1 - 1];

        // Determine if compression is used
        $textData = $this->getTextSectionData($sectionAttributes, $sectionContentOffset);

        // Read verses from the decompressed text data
        $br = new BintexReader($textData, $contentOffset);

        return $this->readVerses($br, $verseCount, $textEncoding);
    }

    /**
     * Get the decompressed text section data.
     * If compressed with snappy-blocks, decompress all blocks.
     */
    private function getTextSectionData(?array $sectionAttributes, int $sectionContentOffset): string
    {
        if ($sectionAttributes === null) {
            // No attributes — uncompressed, return raw content
            $entry = $this->sectionEntries['text'];
            return substr($this->data, $sectionContentOffset, $entry['content_size']);
        }

        $compressionName = $sectionAttributes['compression.name'] ?? null;
        if ($compressionName === null) {
            $entry = $this->sectionEntries['text'];
            return substr($this->data, $sectionContentOffset, $entry['content_size']);
        }

        if ($compressionName !== 'snappy-blocks') {
            throw new RuntimeException("YES2: Unsupported compression: {$compressionName}");
        }

        $compressionInfo = $sectionAttributes['compression.info'] ?? [];
        $blockSize = (int) ($compressionInfo['block_size'] ?? 0);
        $compressedBlockSizes = $compressionInfo['compressed_block_sizes'] ?? [];

        if ($blockSize <= 0 || empty($compressedBlockSizes)) {
            throw new RuntimeException('YES2: Invalid snappy-blocks compression info');
        }

        // Compute offsets for each compressed block
        $offsets = [];
        $cumulative = 0;
        foreach ($compressedBlockSizes as $size) {
            $offsets[] = $cumulative;
            $cumulative += $size;
        }

        // Decompress all blocks
        $decompressed = '';
        foreach ($compressedBlockSizes as $i => $compSize) {
            $blockData = substr($this->data, $sectionContentOffset + $offsets[$i], $compSize);
            $decompressed .= SnappyDecompressor::decompress($blockData);
        }

        return $decompressed;
    }

    /**
     * Read verse texts from the bintex stream.
     *
     * @return string[]
     */
    private function readVerses(BintexReader $br, int $verseCount, int $textEncoding): array
    {
        $verses = [];
        for ($i = 0; $i < $verseCount; $i++) {
            $verseLen = $br->readVarUint();
            $raw = $br->readRaw($verseLen);

            if ($textEncoding === 1) {
                // ASCII / Latin-1
                $verses[] = mb_convert_encoding($raw, 'UTF-8', 'ISO-8859-1');
            } else {
                // UTF-8 (textEncoding == 2, default)
                $verses[] = $raw;
            }
        }
        return $verses;
    }

    // -- Pericopes --

    /**
     * Load pericope (section heading) data.
     *
     * @return array<int, array{ari: int, title: ?string, parallels: string[]}>
     */
    public function loadPericopes(): array
    {
        $versionInfo = $this->getVersionInfo();
        if (($versionInfo['hasPericopes'] ?? 0) === 0) {
            return [];
        }

        $sectionAttributes = $this->readSectionAttributes('pericopes');
        $contentOffset = $this->getSectionContentOffset('pericopes');
        if ($contentOffset === null) {
            return [];
        }

        // Handle compression
        $data = $this->getSectionDecompressedData('pericopes', $sectionAttributes, $contentOffset);
        $br = new BintexReader($data);

        $version = $br->readUint8();
        if ($version !== 2 && $version !== 3) {
            throw new RuntimeException("YES2: Unsupported pericope version: {$version}");
        }

        /* index_size = */ $br->readInt();
        $entryCount = $br->readInt();

        $aris = [];
        $offsets = [];

        if ($version === 2) {
            for ($i = 0; $i < $entryCount; $i++) {
                $aris[] = $br->readInt();
                $offsets[] = $br->readInt();
            }
        } else {
            // version 3 — delta-encoded aris and offsets
            $lastAri = 0;
            $lastOffset = 0;
            for ($i = 0; $i < $entryCount; $i++) {
                $dataAri = $br->readUint16();
                if (($dataAri & 0x8000) === 0) {
                    // absolute — 4 bytes total
                    $ari = ($dataAri << 16) | $br->readUint16();
                } else {
                    // relative delta
                    $ari = $lastAri + ($dataAri & 0x7fff);
                }
                $aris[] = $ari;
                $lastAri = $ari;

                $dataOffset = $br->readUint16();
                $offset = $lastOffset + $dataOffset;
                $offsets[] = $offset;
                $lastOffset = $offset;
            }
        }

        // Data begins after the index
        $dataStartPos = $br->getPos();
        $result = [];

        for ($i = 0; $i < $entryCount; $i++) {
            $br->setPos($dataStartPos + $offsets[$i]);

            $blockVersion = $br->readUint8();
            $title = $br->readValueString();
            $parallelCount = $br->readUint8();
            $parallels = [];
            for ($j = 0; $j < $parallelCount; $j++) {
                $parallels[] = $br->readValueString();
            }

            $result[] = [
                'ari' => $aris[$i],
                'title' => $title,
                'parallels' => $parallels,
            ];
        }

        return $result;
    }

    /**
     * Get pericopes for a specific chapter.
     *
     * @return array<int, array{ari: int, title: ?string, parallels: string[]}>
     */
    public function getPericopesForChapter(int $bookId, int $chapter1): array
    {
        $allPericopes = $this->loadPericopes();
        $result = [];

        foreach ($allPericopes as $pericope) {
            $ari = $pericope['ari'];
            $periBookId = ($ari >> 16) & 0xff;
            $periChapter = ($ari >> 8) & 0xff;

            if ($periBookId === $bookId && $periChapter === $chapter1) {
                $result[] = $pericope;
            }
        }

        return $result;
    }

    // -- Xrefs --

    /**
     * Load cross-reference entries.
     *
     * @return array<int, array{arif: int, content: ?string}>
     */
    public function loadXrefs(): array
    {
        return $this->loadArifSection('xrefs');
    }

    /**
     * Get a specific cross-reference by ARIF.
     */
    public function getXrefEntry(int $arif): ?string
    {
        $xrefs = $this->loadXrefs();
        foreach ($xrefs as $entry) {
            if ($entry['arif'] === $arif) {
                return $entry['content'];
            }
        }
        return null;
    }

    // -- Footnotes --

    /**
     * Load footnote entries.
     *
     * @return array<int, array{arif: int, content: ?string}>
     */
    public function loadFootnotes(): array
    {
        return $this->loadArifSection('footnotes');
    }

    /**
     * Get a specific footnote by ARIF.
     */
    public function getFootnoteEntry(int $arif): ?string
    {
        $footnotes = $this->loadFootnotes();
        foreach ($footnotes as $entry) {
            if ($entry['arif'] === $arif) {
                return $entry['content'];
            }
        }
        return null;
    }

    // -- Generic ARIF section reader (shared by xrefs & footnotes) --

    /**
     * @return array<int, array{arif: int, content: ?string}>
     */
    private function loadArifSection(string $sectionName): array
    {
        $sectionAttributes = $this->readSectionAttributes($sectionName);
        $contentOffset = $this->getSectionContentOffset($sectionName);
        if ($contentOffset === null) {
            return [];
        }

        $data = $this->getSectionDecompressedData($sectionName, $sectionAttributes, $contentOffset);
        $br = new BintexReader($data);

        $version = $br->readUint8();
        if ($version !== 1) {
            throw new RuntimeException("YES2: Unsupported {$sectionName} version: {$version}");
        }

        $entryCount = $br->readInt();

        $arifs = [];
        for ($i = 0; $i < $entryCount; $i++) {
            $arifs[] = $br->readInt();
        }

        $offsets = [];
        for ($i = 0; $i < $entryCount; $i++) {
            $offsets[] = $br->readInt();
        }

        $contentStartPos = $br->getPos();
        $result = [];

        for ($i = 0; $i < $entryCount; $i++) {
            $br->setPos($contentStartPos + $offsets[$i]);
            $content = $br->readValueString();
            $result[] = [
                'arif' => $arifs[$i],
                'content' => $content,
            ];
        }

        return $result;
    }

    /**
     * Decompress section data if it has snappy-blocks compression, otherwise return raw.
     */
    private function getSectionDecompressedData(string $sectionName, ?array $sectionAttributes, int $contentOffset): string
    {
        $entry = $this->sectionEntries[$sectionName] ?? null;
        if ($entry === null) {
            throw new RuntimeException("YES2: Section '{$sectionName}' not found");
        }

        if ($sectionAttributes === null) {
            return substr($this->data, $contentOffset, $entry['content_size']);
        }

        $compressionName = $sectionAttributes['compression.name'] ?? null;
        if ($compressionName === null) {
            return substr($this->data, $contentOffset, $entry['content_size']);
        }

        if ($compressionName !== 'snappy-blocks') {
            throw new RuntimeException("YES2: Unsupported compression: {$compressionName}");
        }

        $compressionInfo = $sectionAttributes['compression.info'] ?? [];
        $blockSize = (int) ($compressionInfo['block_size'] ?? 0);
        $compressedBlockSizes = $compressionInfo['compressed_block_sizes'] ?? [];

        if ($blockSize <= 0 || empty($compressedBlockSizes)) {
            throw new RuntimeException("YES2: Invalid snappy-blocks compression info for section '{$sectionName}'");
        }

        $offsets = [];
        $cumulative = 0;
        foreach ($compressedBlockSizes as $size) {
            $offsets[] = $cumulative;
            $cumulative += $size;
        }

        $decompressed = '';
        foreach ($compressedBlockSizes as $i => $compSize) {
            $blockData = substr($this->data, $contentOffset + $offsets[$i], $compSize);
            $decompressed .= SnappyDecompressor::decompress($blockData);
        }

        return $decompressed;
    }

    // -- ARI helpers --

    /**
     * Encode an ARI from book ID, chapter, and verse.
     */
    public static function encodeAri(int $bookId, int $chapter, int $verse): int
    {
        return (($bookId & 0xff) << 16) | (($chapter & 0xff) << 8) | ($verse & 0xff);
    }

    /**
     * Decode an ARI into [bookId, chapter, verse].
     *
     * @return array{bookId: int, chapter: int, verse: int}
     */
    public static function decodeAri(int $ari): array
    {
        return [
            'bookId' => ($ari >> 16) & 0xff,
            'chapter' => ($ari >> 8) & 0xff,
            'verse' => $ari & 0xff,
        ];
    }

    /**
     * Decode an ARIF into [ari, index].
     *
     * @return array{ari: int, index: int}
     */
    public static function decodeArif(int $arif): array
    {
        return [
            'ari' => ($arif >> 8) & 0xffffff,
            'index' => $arif & 0xff,
        ];
    }
}
